redis/cluster.py

```diff
@@ -0,0 +1,1884 @@
+                 read_from_replicas=False, cluster_error_retry_attempts=3,
+                 debug=False, **kwargs):
+        """
+        """
+        self.command_stack = []
+        self.debug_mode = debug
+        self.nodes_manager = nodes_manager
+        self.refresh_table_asap = False
+        self.result_callbacks = (result_callbacks or
+                                 self.__class__.RESULT_CALLBACKS.copy())
+        self.startup_nodes = startup_nodes if startup_nodes else []
+        self.read_from_replicas = read_from_replicas
+        self.command_flags = self.__class__.COMMAND_FLAGS.copy()
+        self.cluster_response_callbacks = cluster_response_callbacks
+        self.cluster_error_retry_attempts = cluster_error_retry_attempts
+
+        self.encoder = Encoder(
+            kwargs.get("encoding", "utf-8"),
+            kwargs.get("encoding_errors", "strict"),
+            kwargs.get("decode_responses", False),
+        )
+
+        # The commands parser refers to the parent
+        # so that we don't push the COMMAND command
+        # onto the stack
+        self.commands_parser = CommandsParser(super())
+
+    def __repr__(self):
+        """
+        """
+        return "{0}".format(type(self).__name__)
+
+    def __enter__(self):
+        """
+        """
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """
+        """
+        self.reset()
+
+    def __del__(self):
+        try:
+            self.reset()
+        except Exception:
+            pass
+
+    def __len__(self):
+        """
+        """
+        return len(self.command_stack)
+
+    def __nonzero__(self):
+        "Pipeline instances should  always evaluate to True on Python 2.7"
+        return True
+
+    def __bool__(self):
+        "Pipeline instances should  always evaluate to True on Python 3+"
+        return True
+
+    def execute_command(self, *args, **kwargs):
+        """
+        """
+        return self.pipeline_execute_command(*args, **kwargs)
+
+    def pipeline_execute_command(self, *args, **options):
+        """
+        """
+        self.command_stack.append(
+            PipelineCommand(args, options, len(self.command_stack)))
+        return self
+
+    def raise_first_error(self, stack):
+        """
+        """
+        for c in stack:
+            r = c.result
+            if isinstance(r, Exception):
+                self.annotate_exception(r, c.position + 1, c.args)
+                raise r
+
+    def annotate_exception(self, exception, number, command):
+        """
+        """
+        cmd = ' '.join(map(safe_str, command))
+        msg = 'Command # %d (%s) of pipeline caused error: %s' % (
+            number, cmd, exception.args[0])
+        exception.args = (msg,) + exception.args[1:]
+
+    def execute(self, raise_on_error=True):
+        """
+        """
+        stack = self.command_stack
+
+        if not stack:
+            return []
+
+        try:
```

Comment:
Shouldn't send_cluster_commands operate based on the value of the stack... so the if statement above belongs in here?